File: 프로그래머스/lv4/43236. 징검다리/징검다리.kt
class Solution {
    fun solution(distance: Int, rocks: IntArray, n: Int): Int {
        val rock = rocks.sorted()
        var answer = 0
        var left = 1
        var right = distance - rock.first()
        
        while (left <= right) {
            val mid = (left + right) / 2
            var now = 0
            var cnt = 0
            
            for (i in rocks.indices) {
                if (rock[i] - now < mid) {
                    cnt++
                    
                    if (cnt > n) {
                        break
                    }
                } else {
                    now = rock[i]
                }
            }
            
            if (cnt > n) {
                right = mid - 1
            } else {
                answer = mid
                left = mid + 1
            }
        }
        
        return answer
    }
}

File: 프로그래머스/lv4/43236. 징검다리/징검다리.py
def solution(distance, rocks, n):
    rocks.sort()
    answer = 0
    left, right = 1, distance - rocks[0]
    
    while left <= right:
        mid = (left + right) // 2
        cnt, stone = 0, 0
        
        for rock in rocks:
            if rock - stone < mid:
                cnt += 1
            else:
                stone = rock
            
            if cnt > n:
                break
                
        if cnt > n:
            right = mid - 1
        else:
            answer = mid
            left = mid + 1
               
    return answer